### PR TITLE
Expose Prettier configuration

### PR DESCRIPTION
**Pull request by gkze**
* Expose Prettier configuration options, taken from
  http://json.schemastore.org/prettierrc
* Wrap Prettier invocation in a command that passes generated
  configuration file from passed options

### CODE REVIEW

**Comment by gkze** [comment]
Stacking on top of #102 to clear `nix flake check` warnings, happy to split

File: programs/prettier.nix [hunk 1]
```diff
@@ -0,0 +1,282 @@
+{ lib, pkgs, config, ... }:
+let
+  inherit (lib) filterAttrsRecursive literalMD mkEnableOption mkOption types;
+
+  cfg = config.programs.prettier;
+  configFormat = pkgs.formats.json { };
+
+  # Configuration scheam for prettier, which we generate prettierrc.json with.
```
Comment (zimbatm) [inline programs/prettier.nix:8]: s/scheam/schema/

File: programs/prettier.nix [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+      description = "Provide a list of patterns to override prettier configuration.";
+      type = types.nullOr (types.listOf types.attrs);
+      example = { files = [ "*.html" "legacy/**/*.js" ]; options.tabwidth = 4; };
+      default = null;
+    };
+  };
+
+  # We filter out null values in dprint configuration since dprint errors when
+  # they're present
+  mkConfigFile = { filename ? ".prettierrc.json", config ? { } }:
```
Comment (zimbatm) [inline programs/prettier.nix:210]: let's use `"prettierrc.json"`. That argument can also be collapsed as it's not used or exposed anywhere.

File: programs/prettier.nix [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      type = types.package;
+      default = pkgs.nodePackages.prettier;
+      defaultText = lib.literalExpression "pkgs.nodePackages.prettier";
+      description = lib.mdDoc ''
+        prettier derivation to use.
+      '';
+    };
+
+    # Represents the .prettierrc.json config schema
+    config = configSchema;
```
Comment (zimbatm) [inline programs/prettier.nix:227]: "config" is an overloaded term in the nix module system; it's best to use "settings" instead.

File: programs/prettier.nix [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    excludes = mkOption {
+      description = "Path / file patterns to exclude for Prettier";
+      type = types.listOf types.str;
+      default = [ ];
+    };
+
+    # Wrapped prettier invocation passing a generated .prettierrc.json
+    # configuration
+    wrapper = mkOption {
+      description = "The prettier package, wrapped with the config file.";
```
Comment (zimbatm) [inline programs/prettier.nix:259]: this option should be marked as private as it's not meant to be used directly by users

File: programs/prettier.nix [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+      defaultText = literalMD "wrapped `prettier` command";
+      default =
+        let
+          configFile = mkConfigFile { config = cfg.config; };
+          x = pkgs.writeShellScriptBin "prettier" ''
+            set -euo pipefail
+            exec ${lib.getBin cfg.package}/bin/prettier --config=${configFile} "$@"
+          '';
+        in
+        (x // { meta = config.package.meta // x.meta; });
```
Comment (zimbatm) [inline programs/prettier.nix:270]: How would that interact with users that have a prettierrs.json in their repo? Maybe not wrap the program if the user is not adding any config in nix?

Comment (gkze) [reply programs/prettier.nix:308]: Good point, I added a conditional in the wrapper to check for settings being empty and using prettier directly if so. Let me know if there is a better way.

**Comment by zimbatm** [comment]
looks good. Can you just squash and rebase to remove the other changes from the PR?

**Comment by gkze** [comment]
Lmk if that works

**Comment by zimbatm** [comment]
thanks. I removed the wrapper as it wasn't necessary in the end. Maybe it's worth applying the same changes to drpint.

**Comment by gkze** [comment]
👍🏻 https://github.com/numtide/treefmt-nix/pull/105

File: programs/prettier.nix [hunk 6]
```diff
@@ -1,25 +1,275 @@
...
+      default = null;
+    };
+    plugins = mkOption {
+      description = ''
+        Add a plugin. Multiple plugins can be passed as separate `--plugin`s.
+      '';
+      type = types.nullOr (types.listOf types.str);
+      example = [ "@prettier/plugin-xml" ];
+      default = null;
+    };
```
Comment (srid) [inline programs/prettier.nix:143]: https://github.com/numtide/treefmt-nix/issues/112

**Comment by drupol** [comment]
Hello,

Today I just noticed that using `proseWrap` does not have any single effect. However, to get it working properly I added:

```nix
 prettier = {
   options = [
     "--prose-wrap=always"
   ];
   editorconfig = true;
};
```

**Pull request merged**